src/gluonnlp/data/transforms.py

```diff
@@ -988,10 +1009,246 @@ def convert_tokens_to_ids(self, tokens):
+            # length is less than the specified length.
+            # Account for [CLS], [SEP], [SEP] with "- 3"
+            self._truncate_seq_pair(tokens_a, tokens_b, self._max_seq_length - 3)
+        else:
+            # Account for [CLS] and [SEP] with "- 2"
+            if len(tokens_a) > self._max_seq_length - 2:
+                tokens_a = tokens_a[0:(self._max_seq_length - 2)]
+
+        # The embedding vectors for `type=0` and `type=1` were learned during
+        # pre-training and are added to the wordpiece embedding vector
+        # (and position vector). This is not *strictly* necessary since
+        # the [SEP] token unambiguously separates the sequences, but it makes
+        # it easier for the model to learn the concept of sequences.
+
+        # For classification tasks, the first vector (corresponding to [CLS]) is
+        # used as as the "sentence vector". Note that this only makes sense because
+        # the entire model is fine-tuned.
+        vocab = self._tokenizer.vocab
+        tokens = []
+        segment_ids = []
+        tokens.append(vocab.cls_token)
+        segment_ids.append(0)
+        for token in tokens_a:
+            tokens.append(token)
+            segment_ids.append(0)
+        tokens.append(vocab.sep_token)
+        segment_ids.append(0)
+
+        if tokens_b:
+            for token in tokens_b:
+                tokens.append(token)
+                segment_ids.append(1)
+            tokens.append(vocab.sep_token)
+            segment_ids.append(1)
+
+        input_ids = self._tokenizer.convert_tokens_to_ids(tokens)
+
+        # The valid length of sentences. Only real  tokens are attended to.
+        valid_length = len(input_ids)
+
+        if self._pad:
+            # Zero-pad up to the sequence length.
+            padding_length = self._max_seq_length - valid_length
+            # use padding tokens for the rest
+            input_ids.extend([vocab[vocab.padding_token]] * padding_length)
+            segment_ids.extend([0] * padding_length)
+
+        return np.array(input_ids, dtype='int32'), np.array(valid_length, dtype='int32'),\
+            np.array(segment_ids, dtype='int32')
+
+    def _truncate_seq_pair(self, tokens_a, tokens_b, max_length):
+        """Truncates a sequence pair in place to the maximum length."""
+        # This is a simple heuristic which will always truncate the longer sequence
+        # one token at a time. This makes more sense than truncating an equal percent
+        # of tokens from each, since if one sequence is very short then each token
+        # that's truncated likely contains more information than a longer sequence.
+        while True:
+            total_length = len(tokens_a) + len(tokens_b)
+            if total_length <= max_length:
+                break
+            if len(tokens_a) > len(tokens_b):
+                tokens_a.pop()
+            else:
+                tokens_b.pop()
+
+
+class BERTDatasetTransform(object):
```

Comment:
On second thought this is very task specific transformation code, compared to `BERTSentenceTransform `. `BERTSentenceTransform` can be generally used as a building block for many tasks like regression, classification and QA. Would it be better to put this transform back to the /script/bert folder? Here's my thought:

Instead, we should need to build composable transformation pipelines:

Currently the `Dataset` API supports `dataset.transform(t1).transform(t2) ... `, where both `t1` and `t2` transform all fields in the input. However, sometimes we want to reuse some `Transform` API for a subset of the fields in the input. For example, say we have a `Dataset` of 3 fields: text_a, text_b and label. A desirable API should allow:
```
trans_fn = ComposableTransform([Tokenizer(), Tokenizer(), LabelTransform()])
dataset.transform(trans_fn)
```
here the 3 fields in a data sample are transformed with 3 transforms respectively. If this is reasonable we can consider doing this in future PRs

@kenjewu @leezu @szha what do you guys think?

Reply:
batchify has an interface for a similar purpose, called [Tuple](http://gluon-nlp.mxnet.io/master/api/modules/data.batchify.html#gluonnlp.data.batchify.Tuple).

Reply:
It's probably a good idea to have the same for transform.

Reply:
Note that batchify function deals with a different signature, so for simple sample-wise transforms, the tuple function may need to be different.

Reply:
Ok, thank you very much for your comments.